Xenobiology bluespace crossbreeds booster pack

Caldony opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Adds nearly all of the slime bluespace crossbreeds. These crossbreeds are capable of drawing runes on the floor with different effects and be stored back in the crossbreed that made them to be reused elsewhere. The most powerful crossbreeds will usually take more time to draw.

Bluespace runes can now be cleaned normally. Although they can be destroyed much faster with a bluespace crystal.

The most notable(unbalanced) crossbreeds are :
warping red : give higher punch damage to whoever is standing on the rune, gloves of the north star compatible
warping sepia : basically a reusable timestop trap, except it's very much visible.
warping gold : turn anything over the rune into gold every 30 seconds. only one item at a time.
warping purple : transmutes plastic and cloth into suture and regenerative mesh.
warping light pink : will admin heal a soulless corpse if it finds a willing ghost.
warping rainbow : teleports you to your own personal room! like a really shitty hilbert's hotel because I can't map.
and so much more!

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Slime crossbreeds haven't been updated in forever, and bluespace crossbreeds are not only a unique addition to the xenobiologist arsenal, but they can also help the crew passively in a lot of different ways.

## Changelog
:cl:
add: xeno resin sheets. Used to build already hatched xeno eggs, resin seeds, resin walls and resin nest (only buildable above resin seeds/node). Mainly to make the crew freak out for fake xenomorphs.

add: Nearly all of the bluespace crossbreeds are now available. Experiment!

/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Caldony committed:
Creates the bluespace crossbreed

Adds the bluespace crossbreed (not complete)

Caldony committed:
Update xenobiology.dm

Caldony committed:
Update warping

Caldony committed:
Update warping

Caldony committed:
Delete slimecrossing.dmi

Caldony committed:
Add files via upload

Caldony committed:
fix two color_slime variable missing quotes

TiviPlus commented:
I've written essays shorter than that changelog

Caldony committed:
remove a useless process()

ArcaneMusic commented:
It's perfectly alright to say "Bluespace slime crossbreeds are now available, experiment!"

caluany committed:
Revert "remove a useless process()"

This reverts commit c7fe3adfddde501bfc43350c31083f500a413056.

optimumtact commented:
Why cant' they be cleaned?

Caldony commented:
@optimumtact  It doesn't feel right to have these runes be as easily removable as cult runes considering they are harder to produce and take much more time to draw for the actually harmful runes. I do plan on adding a few more ways to delete them so they are not impossible to counter if the miners aren't doing their job though.

optimumtact commented:
I don't think that's a justification for them to not be cleanable honestly

Caldony commented:
@optimumtact after further consideration I think you'r right, people won't necessarily know that they are cleanable in a different way, so I'll just make it faster to clean it with a bluespace crystal but still cleanable with a mop like cult runes. It does make more sense ultimately.

Caldony committed:
Fix a few bug and make the rune cleanable.

The rune can now be cleaned by soap and mops. although using a bluespace crystal is much faster.
Fixed a few bugs where someone could store/draw the rune to get around the cooldown

Caldony committed:
Polishing and balance.

replace most "get_turf(src)" with T. Fix a few fluff descriptions.Adds fluff text when you draw the rune. Remove a few TODO comments for things I already finished.Fixes a few description and rebalance a few runes drawing time such as the red rune.

Caldony committed:
Rebalances the red rune

The red rune is compatible with the gloves of the north star and would always stun. The rune has been rebalanced so that each punch in a row on the rune will deal stamina damage to the puncher and lower their punch damage. 

The punch damage can be reset by exiting and re-entering the rune. It has a bug where the unarmed_attack signal triggers whenever the puncher click on anything close to him with an empty hand.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/effect/warped_rune
+	name = "warped rune"
+	desc = "An unstable rune born of the depth of bluespace"
```
*depths

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+			qdel(BC)
+///destroyable by xenomorphs			
+/obj/effect/warped_rune/acid_act() 
+	visible_message("<span class='warning'>The rune is being disolved!</span>")
```
*dissolved

also, if the rune is dissolved instantly by acid, why are you giving a message that it's "being" dissolved (implying that it's still dissolving)?

also, you might wanna use "[src]" instead of "the rune"

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+//nearly all runes use their turf in some way so we set T to their turf automatically.
+/obj/effect/warped_rune/Initialize()
+	..()
+	T = get_turf(src)
```
and what happens if the rune gets moved somehow? I think it'd be much better to just use get_turf(src) whenever you'd normally use T

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+		to_chat(user, "<span class='warning'>There is already a bluespace rune here!</span>")
+		return
+	if(istype(target, runepath))       //checks if the target is a rune and then if you can store it
+		if(warp_charge == max_charge)
```
use if(warp_charge >= max_charge) here instead

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	if(warp_charge > 0 && do_after(user, drawing_time,target = target)) //finally spawns the right rune if you have charge(s) left
+		playsound(target, 'sound/effects/slosh.ogg', 20, TRUE)
+
+		if(warp_charge == 0) //this is only here to fix a bug where the user can draw multiple runes at once if they spam click
```
use warp_charge <= 0 instead of warp_charge==0

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+			to_chat(user, "<span class='warning'>Your crossbreed is already full!</span>")
+			return
+		if(do_after(user, storing_time,target = target) && warp_charge < max_charge)
+			to_chat(user, "<span class='notice'>You store the rune in your bluespace crossbreed</span>")
```
use "[src]." (don't forget the period) instead of "your bluespace crossbreed"

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+		playsound(target, 'sound/effects/slosh.ogg', 20, TRUE)
+
+		if(warp_charge == 0) //this is only here to fix a bug where the user can draw multiple runes at once if they spam click
+			to_chat(user, "<span class='warning'>Your crossbreed is empty!</span>")
```
use "[src]" instead of "Your crossbreed"

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+			return
+		warp_charge--
+		new runepath(target)
+		to_chat(user, "<span class='notice'>You carefully draw the rune with your crossbreed.</span>")
```
use "[src]" instead of "your crossbreed"

I think you get the picture here.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/item/slimecross/warping/grey
+	colour = "grey"
+	effect_desc = "Creates a rune.Extracts the rune is on are absorbed, 8 extracts produces a slime of that color."
```
forgot a space after that period

ATH1909 replied:
just say "Extracts that are on the rune are absorbed" instead of "Extracts the rune is on are absorbed"

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+		slimerevival()
+	
+///Makes a slime of the color of extract that were put on the runes.can only take one type of extract between slime spawning.
+/obj/effect/warped_rune/greyspace/proc/slimerevival()
```
this rune is kind of stupidly inefficient. when would I ever use this?

IMO, you should make this rune instead convert ONE slime core and like 8 monkey cubes into a baby slime of that slime core's color. If the slime core was spent, the rune recharges it (to 1 charge) instead of making a baby slime from it. _That_ would be a rune that would at least see _some_ use. Remember, people are spending 11 slime cores to make each of these runes.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	RegisterSignal(T,COMSIG_ATOM_ENTERED,.proc/teleport_fire) //TODO : figure out how to use the shortened version of this
+	START_PROCESSING(SSprocessing, src)
+
+/obj/effect/warped_rune/orangespace/process()
```
make this happen in Crossed() or something, not process()

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/*
+		for(F.fire_stacks > 0 && !istype(F,/mob/living/simple_animal)) //Keeps teleporting the guy as long as he is on fire (doesn't include simplemobs) WIP
+			do_teleport(F, get_turf(F), 5, asoundin = 'sound/effects/phasein.ogg', channel = TELEPORT_CHANNEL_BLUESPACE)
```
there's no delay for this, so people who get teleported by this rune are gonna get teleported like 10+ times per second

also, this rune won't set anyone else who was on the rune on fire or teleport them until the people before them in the queue have been extinguished.

use, like, callbacks or something for this

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/effect/warped_rune/orangespace
+	name = "orange space rune"
+	desc = "When all is reduced to ash, it shall be reborn from the depth of bluespace"
```
period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+
+
+
```
alright, that's Part I of this review, I'll get to Part II later

ATH1909 commented:
![image](https://user-images.githubusercontent.com/42606352/77471980-3572b300-6de1-11ea-9b17-1bc7f51fb696.png)

oof

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	var/obj/item/stack/medical/mesh/M            //M is for mesh and you
+	max_cooldown = 30
+
+/obj/effect/warped_rune/purplespace/Initialize()
```
welcome back to part 2 of my ~~meme~~ PR review

this rune (and all of the others that start processing, including the ones before this one) needs to stop processing once it's destroyed

ATH1909 replied:
welcome back to part 2 of my meme PR review

this rune (and all of the others that start processing, including the ones before this one) needs to stop processing once it's destroyed

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+///""transforms"" cloth and plastic into suture and regenerative mesh
+/obj/effect/warped_rune/purplespace/proc/transmute_heal() //transmute plastic and cloth into suture and regenerative mesh
+	for(var/obj/item/stack/sheet/plastic/P in T)  //transmute Plastic into regenerative mesh
```
you might wanna check that there's at least 1 sheet in these plastic and cloth stacks? idk, it's a pretty unlikely edge case scenario, but you never know when some bug somewhere is gonna let you make a cloth or plastic stack with 0 sheets in it.

you also might want to make the ratio of plastic or cloth to mesh or sutures be 2:1 instead of 1:1, since this rune is a bit TOO efficient right now, IMO (and meshes and sutures only stack up to 25 "sheets" per stack anyway, so it works out).

still, I really like the idea of this rune, and I think that this could be something that people might actually specifically ask xenobiologists to make for them, sort of like how roboticists and borgs specifically request yellow slime core cells.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+
+
+/* the blue warp rune  keeps a tile slippery CONSTANTLY by adding water (maybe I'll make it lube if I feel extra fucky) every few(10) seconds */
```
yeah, make it lube the tile. you're spending 11 slime cores on this rune, you should get your damn money's worth (plus, a constantly lubed tile could get people to slip onto other runes, which could make for some hilarious rube goldberg-esque traps)

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+		slippery_rune(T)
+
+
+///just wet the turf under the rune every now and then, there's actually a small moment where the tile isn't wet
```
bruh imagine making a rune whose literal ONLY PURPOSE is to keep a tile continuously W E T and it can't even do that smh my head my head

seriously, though, make this actually make a tile always slippery

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	START_PROCESSING(SSprocessing, src)
+
+/obj/effect/warped_rune/cyanspace/process()
+	if(cooldown < world.time)
```
does this rune even need a cooldown?

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/item/slimecross/warping/metal
+	colour = "metal"
+	runepath = /obj/effect/warped_rune/metalspace
+	effect_desc = "Creates a rune that prevents passage above it, takes longer to store and draw than other runes."
```
this is a pretty shitty rune, not gonna lie. can you at least make this rune, like, nigh-invisible (sort of like ninja stealth)? or make it possible for certain people (like, say, only slimepeople and slimes?) to walk through this barrier without being impeded?

ATH1909 replied:
mm, actually, there's no way to remove this rune without acid or soap or something, so I guess making a GBJ with 4 of these is a pretty decent use for them (until a janitor casually removes your 40 bluespace slime cores and 4 metal slime cores's worth of runes), although I'd still recommend buffing them in one of the ways I described above (because making GBJs is a pretty niche purpose for metal slime runes)

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/effect/warped_rune/cyanspace
+	icon_state = "slipperyspace"
+	desc = "Water drowns, water feeds, water slips, is there a thing water cannot do?"
```
might wanna turn that comma after slips into a period or a semicolon or something, idk (it might be fine as it is now)

ATH1909 replied:
also, replace "a thing" with "anything"

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+//It's a wall what do you want from me
+/obj/effect/warped_rune/metalspace
+	desc = "Words are powerful things, they can stop someone dead in their tracks if used in the right way"
```
need a period at the end of this

also, I'm not sure if that comma should be replaced with another kind of punctuation mark. commas were never really my strong point, so I think I'm gonna mostly ignore them from here on out in this review, since I don't _really_ know what I'm talking about as well as I should when it comes to commas.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/item/slimecross/warping/yellow
+	colour = "yellow"
+	runepath = /obj/effect/warped_rune/yellowspace
+	effect_desc = "Creates a rune that Infinitely recharge batteries that are put on it, it will also passively recharge the APC of the room"
```
"Infinitely" shouldn't be capitalized

"the room" sounds a bit odd. maybe replace it with "the room it's in."?

you need a period at the end of this sentence

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+
+/obj/effect/warped_rune/yellowspace
+	desc = "Where does all this energy come from? Who knows,the process does not matter, only the result."
```
need a space after that first comma

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	if(cooldown < world.time)
+		cooldown = world.time + max_cooldown
+		apc_charge()
+		electrivore()
```
excuse me what the fuck

this isn't even vore, smh my head my head

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+			return
+		if(P.cell.charge <! (P.cell.maxcharge - 1))
+			return
+		P.cell.charge += P.cell.maxcharge/100 //will basically recharge 1% of the APC cell every 10 seconds,we don't want to run engineers out of buisness
```
this APC-recharging stuff needs a buff in recharging speed. remember that you could have just slapped a yellow slime cell into an APC to make that APC basically unable to run out of power anyway, so the only advantage that making a rune out of your yellow slime instead of a yellow slime cell would have is the ability to recharge other cells and the portability/ease of setting it up, and the yellow slime cell is 10 bluespace slime cores cheaper than this rune is

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/effect/warped_rune/yellowspace/proc/electrivore()
+	for(var/obj/item/stock_parts/cell/C in T)
+		if(C.charge < C.maxcharge) //check if the battery is not full
+			C.charge += C.maxcharge/10
```
if a cell is at 99% charge, this can bring it to 109% charge. you need to make this a clamp or something

also, if you could make this recharge cells that are inside of guns and defibs and borgs and stuff too, that'd be swell (and make this rune much more useful for, say, fighting blobs (and just much more useful in general, really))

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+
+
+
```
and that concludes Part II of my review

Part III is coming soon

Caldony replied:
Thanks! I'll try to fix most of the issues you've talked about. I know it's probably a lot of work reviewing all the crossbreed so I appreciate it.

ATH1909 replied:
Nah, I haven't done nearly as much work as you have on this PR.

Despite my many comments, I really do like what you've done here in this PR, and I think that the runes presented in this PR can perhaps make up for the massive loss of fun in the xenobio department that came with the removal of change carps (from xenobio).

cacogen commented:
those sprites

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	effect_desc = "Makes a rune that will periodically create pure plasma, if grinded it can produce a lot of plasma dust but that's it"
+
+
+/obj/effect/warped_rune/darkpurplespace
```
alright, let's start Part III of my PR review

Instead of having this rune make plasma crystals, why not just have it add the plasma reagent directly to the contents of a random reagent container that's on its tile every time it processes? That way, these runes won't end up creating way more plasma crystals than any sane xenobiologist would reasonably need, and you don't need to micromanage a reagent grinder to make proper use of them.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/item/slimecross/warping/darkpurple
+	colour = "dark purple"
+	runepath = /obj/effect/warped_rune/darkpurplespace
+	effect_desc = "Makes a rune that will periodically create pure plasma, if grinded it can produce a lot of plasma dust but that's it"
```
period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	var/obj/effect/overlay/holotile
+
+
+/obj/effect/warped_rune/ceruleanspace/Initialize()
```
used Crossed() instead of COSMIG_ATOM_ENTERED

that way, you don't have to use a for loop to go through every mob on the tile, you can just specifically copy the person who just entered the rune

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/item/slimecross/warping/pyrite
+	colour = "pyrite"
+	runepath = /obj/effect/warped_rune/pyritespace
+	effect_desc = "draws a rune that will randomly color whoever steps on it"
```
capitalize "draws"

period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	effect_desc = "draws a rune that will randomly color whoever steps on it"
+
+/obj/effect/warped_rune/pyritespace
+	desc = "Who shall we be today? they asked, but not even the canvas would answer."
```
Add a ' before "Who" and after the question mark.

I also think you need a comma after the ' that goes after the question mark, but I'm not sure (I've never been good with quotation mark puntuation rules). I could be wrong, though.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	desc = "Who shall we be today? they asked, but not even the canvas would answer."
+	icon_state = "colorune"
+
+/obj/effect/warped_rune/pyritespace/Initialize()
```
use Crossed()

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	RegisterSignal(T,COMSIG_ATOM_ENTERED,.proc/colored)
+
+///colors whoever steps on the rune randomly
+/obj/effect/warped_rune/pyritespace/proc/colored()
```
not gonna lie, this kinda seems like a filler rune.

then again, pyrite is basically a filler extract/slime type, so 🤷‍♂ . this is fine, I think (at least you can use it to prank people who walk over it in the halls or something).

wait, ooh, idea here: what if we could use this rune to drain the color from things? like, make them monochrome if they stand on the rune for 4 seconds or so. perhaps add some byproduct (that can be used to make something cool?) to this color drainage reaction when it's used on a sapient, non-monochrome creature, so that you can kidnap people to drain them of their color?

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/item/slimecross/warping/red
+	colour = "red"
+	runepath = /obj/effect/warped_rune/redspace
+	effect_desc = "draw a rune severely increasing your punch damage. Only works on one person at a time."
```
*Draws

*that greatly increases

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+    drawing_time = 100 //longer to draw so you don't draw it in the middle of a fight.
+
+/obj/effect/warped_rune/redspace
+	desc = "Progress is made through adversity, power is obtained through violence"
```
period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/effect/warped_rune/redspace/proc/angry_in()
+
+	if(rage) //don't buff anyone if someone is already buffed by the rune.
```
shouldn't this be inside of the for loop?

honestly, why not just make this work on multiple people at the same time? you can move some of this stuff into a status effect so that the original punch damage thresholds don't have to be stored in this rune. or you can just not bother fiddling with punch damage at all and instead use a martial art of some kind.

proposal for a new effect for this rune: teaches anyone who crosses it the psychotic brawling martial art (and colors them red) until they're doused with water.

proposal for a much easier to implement alternative new effect for this rune: adds bath salts to the system of anyone who crosses it (or who stays on it).

Caldony replied:
The only reason I didn't make it work on multiple people at the same time is so that I could redistribute the right punch stats to the person that is leaving instead of the stats of someone else. Although I probably could've worked around that with some list fuckery

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+///triggers whenever the enraged player punches something. lowers their punch damage and deals stamina damage to themselves.
+/obj/effect/warped_rune/redspace/proc/blind_rage()
+	H.dna.species.punchdamagelow -= 10
```
hm yes today I will make my punchdamagelow be a negative number

honestly, this stuff isn't really needed, seeing as how you can only really punch people who are stupid enough to get close to the rune while you're on it and you lose your powers if you step off of the rune. you can also just get around this downside (well, apart from the stamina damage) by stepping off of and then back onto the rune. also, for a proc named blind_rage(), I was kind of expecting a blindness or a blurriness debuff.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+//list of things you can build with resin sheets.
+GLOBAL_LIST_INIT(resin_recipes, list ( \
+	new/datum/stack_recipe("Resin seed", /obj/structure/alien/weeds/node, 10, one_per_turf = 1, on_floor = 1, time = 100), \
```
inconsistent capitalization in the names of these recipes

ATH1909 replied:
also, maybe you could consider allowing people to craft resin into certain xeno organs and/or xeno hides? like, don't let people craft egg sacs or neurotoxin spitters, but the other xeno organs (resin spinners, xeno tongues, hive nodes, xeno eyes, xeno livers, etc.) should be fine. if you think that it doesn't make sense that resin should be allowed to be made into that stuff, consider maybe adding a new xeno material for if you put leather, bones, sinew, hides, or other semi-rare organic material on this rune.

ATH1909 replied:
also, you should add resin windows and resin nests to this list, since those are also structures that xenos can make

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/item/slimecross/warping/pink
+	colour = "pink"
+	effect_desc = "Draw a rune that hugs whoever steps on it.Only useful to friendless people."
```
add a space after the period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	runepath = /obj/effect/warped_rune/pinkspace
+
+/obj/effect/warped_rune/pinkspace
+	desc = "Love is the only reliable source of happiness we have left.But like everything, it comes with a price."
```
add a space after the period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	var/mob/living/carbon/human/loved
+
+
+/obj/effect/warped_rune/pinkspace/Initialize()
```
use Crossed()

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/effect/warped_rune/pinkspace/proc/hug_rune()
+	for(var/mob/living/carbon/human/H in T)
+		playsound(T, "sound/weapons/thudswoosh.ogg", 50, TRUE)
+		SEND_SIGNAL(H, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug)
```
make this give the besthug moodlet, you're spending 10 bluespace slime cores and a pink slime on this thing

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	for(var/mob/living/carbon/human/H in T)
+		playsound(T, "sound/weapons/thudswoosh.ogg", 50, TRUE)
+		SEND_SIGNAL(H, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug)
+		H.visible_message(H, "<span class='notice'>You feel a magical embrace</span>")
```
period

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/item/slimecross/warping/gold
+	colour = "gold"
+	runepath = /obj/effect/warped_rune/goldspace
+	effect_desc = "draw a rune capable of turning nearly anything put on it to gold."
```
*Draws

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/effect/warped_rune/goldspace
+	icon_state = "midas_rune"
+	desc = "When everything is abundant,it becomes easy to forget when one had to work to obtain anything."
```
space after the comma

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	RegisterSignal(T,COMSIG_ATOM_ENTERED,.proc/midas_touch)
+
+///turn anything on the tile to a golden version of themselves. Sheets of any kind will be directly turned into gold bars.
+/obj/effect/warped_rune/goldspace/proc/midas_touch()
```
consider reusing some of the code from the metalgen chem for this

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+		if(istype(S,/obj/item/stack/sheet/mineral/gold)) //can't turn gold into gold
+			return
+		var/obj/item/stack/sheet/mineral/gold/G = new(T)
+		G.amount = rand(0, S.amount) //will give a random amount of gold between 0 and the amount of the original sheets
```
don't make this RNG dependent

plus, the rune is supposed to be turning stuff into gold, not annihilating some of it and turning the rest into gold

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	for(var/obj/item/I in T)
+		I.material_flags = MATERIAL_COLOR | MATERIAL_ADD_PREFIX | MATERIAL_AFFECT_STATISTICS
+		I.set_custom_materials(list(/datum/material/gold=500)) 
+
```
honestly, the effect of this gold rune probably should have been the effect of the metal rune. despite their name, gold slimes are associated with summoning mobs, not the gold material. gold is a metal, so it falls under the domain of metal slimes. I was expecting a gold rune to summon demons and shit in a satanic ritual, not turn stuff into gold.

suggested replacement effect: If a unique station pet (not parrots from a parrot crate, holo pets from the holodeck, or shit like that, I mean an ACTUAL unique station pet, like Ian, Poly, Runtime, Citrus, Tom, etc.) dies upon a gold rune (or has its corpse dragged onto the rune), the rune consumes the pet's corpse, creates a netherworld link (https://github.com/tgstation/tgstation/blob/master/code/modules/mob/living/simple_animal/hostile/netherworld.dm) with a nerfed mob cap (say, 5 mobs) on its tile (if there isn't already a netherworld link there), then deletes itself.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	colour = "adamantine"
+	runepath = /obj/effect/warped_rune/adamantinespace
+
+/obj/item/slimecross/warping/rainbow
```
suggestion: make this like the bluespace rune, but instead of swapping locations, it swaps (or transfers, if only one of the two mobs is sapient) minds (sort of like the recoil effect of a spent book of mind transfer).

ATH1909 replied:
might need to restrict this to organic mobs only, just to prevent cheese with pAIs, minebot intelligence upgrades, posibrains, etc. alternatively, just cut out the "(or transfers, if only one of the two mobs is sapient)" bit.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	colour = "light pink"
+	runepath = /obj/effect/warped_rune/lightpinkspace
+
+/obj/item/slimecross/warping/adamantine
```
suggested effect: frees any golems that are dragged onto the rune

alternative suggested effect: enslaves any golems that are dragged onto the rune to the will of the person who placed the rune. any non-golems who pass over the rune get given adamantine vocal cords.

alternative alternative suggested effect: turns anyone who remains on the rune for more than 10 seconds into a adamantine free golem.

alternative alternative alternative suggested effect: turns anyone who remains on the rune for more than 30 seconds into an adamantine **servant** golem (whose master is the person who placed the rune). I'm not sure if this effect would be too strong, but damn would it be fun to use.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/obj/effect/warped_rune/ceruleanspace/Destroy()
+	..() //needed so the rune still gets stored by the crossbreed
+	qdel(holotile)
+
```
this is a cool effect and all, but what about:
alternative effect: after a (not beheaded) corpse or a head remains on this rune for 5 seconds, it will begin to recite everything it said over the course of the round. whether or not this is in reverse order is up to you, Caldony. you can look at the code for the mind reader mutation and for changelings learning their absorb victims' recent spoken phrases.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+	colour = "black"
+	runepath = /obj/effect/warped_rune/blackspace
+
+/obj/item/slimecross/warping/lightpink
```
proposed effect: dragging the corpse of a suicider onto this rune offers their body up to ghosts. if a ghost chooses to inhabit it, the body is rezzed and fully healed (that includes organ damage so that it won't die immediately upon returning to life).

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+
+/obj/item/slimecross/warping/black
+	colour = "black"
+	runepath = /obj/effect/warped_rune/blackspace
```
proposed effect: standing on this rune for 60 seconds turns you into a pickle.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+/*
+Anything after this is in """progress"""
+
+/*Any explosion over the rune will convert anything in the range of the explosion to SLIME, not people because even I'm not that evil */
```
alternative effect: once every minute (or maybe once every 30 seconds?), this rune inserts a microbomb implant into an empty implant case that's on its tile.

Caldony commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,796 @@
+		for(C in T)
+			count++
+			return
+	for(M in T)   //teleports the first person to the rune and the second person to the old location of the first person, prone to lag if abused
```
Considering monkeys aren't teleportable by the rune (it only works with humans and its subtypes) I doubt it would be very doable. the last time I did get to make it lag is by basically getting another human against the wall and continuously teleporting each other back and forth by just walking in. Although I never managed to reproduce that properly after a few code tweaks.However if you did manage to get a lot of humanised monkey on the same tile you could probably deal a substantial amount of lag.Which is probably what makes this rune one of the most dangerous to have merged as of now.

Caldony committed:
Update warping

reworks a bunch of previously established rune and adds the adamantine rune
silver rune will now absorb nutrients directly from the food put on it without directly destroying the food itself and then feed back whoever is on the tile with nutriments.
purple rune now costs 2 cloth/plastic for suture/regen mesh
grey rune now creates an adult slime of said color instead of a baby slime to facilitate crossbreed production
metal crossbreed now has 4 charges instead of 1 
bluespace rune now no longer is lag abusable because of a cooldown.
adamantine is a rune capable of creating ores depending on the mineral rocks surrounding it such as lavaland ore rocks.
red rune will now only slightly lower your stamina instead of lowering your punch damage over time.
dark purple rune will now directly fill whatever beaker is above it with plasma dust over time.
pink rune now gives a better mood buff (although not a hug related one as of now because of a runtime problem)
grammar error fixing (I probably missed out a few sorry I don't English very good) 
That should probably be all.

Caldony commented:
the title of the update was meant to be "huge rune rework + adamantine" but it bugged out so now we're stuck with just "update warping"

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,865 @@
+		auto_mining()
+
+/obj/effect/warped_rune/adamantinespace/proc/auto_mining()
+	for(var/turf/closed/mineral/M in range(7,T)) //the range is pretty big to at least try to rival miners and their plasma cutters.
```
what does this range have to do with rivaling miners and their plasma cutters? if anything, you'd want a SMALLER range, so that you can more accurately restrict which kinds of ores this rune will be copying from (if the only ore in range is, say, diamond ore, then this rune will produce some chunks of diamond ore every 30 seconds)

Caldony replied:
ho no you don't get it, it will copy ALL ores within it's range every 30 seconds, so the more range it has the more ores at the same time it can get in general.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,865 @@
+
+/obj/effect/warped_rune/adamantinespace/proc/auto_mining()
+	for(var/turf/closed/mineral/M in range(7,T)) //the range is pretty big to at least try to rival miners and their plasma cutters.
+		if(M.mineralType == null) //here to counter runtimes when the mineral type of the rock is null, does absolutely nothing.
```
can't you just use if(M.mineralType!=null) here and remove the else?

Caldony replied:
not really, if we don't use else it will use the next code in line, aka the one we're trying to not use, and using an early return will stop the rune BEFORE it gets to duplicate every rune in range.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,865 @@
+/obj/item/slimecross/warping/adamantine
+	colour = "adamantine"
+	runepath = /obj/effect/warped_rune/adamantinespace
+	effect_desc = "draws a rune capable of copying the ores of nearby mineral rocks."
```
*Draws

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,865 @@
+	effect_desc = "draws a rune capable of copying the ores of nearby mineral rocks."
+
+/obj/effect/warped_rune/adamantinespace  //doesn't have a rune icon yet please spriters help me I can't sprite for shit I beg you
+	desc = "The universe's ressource are nothing but tools for us to use and abuse."
```
*resources

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping:
```diff
@@ -0,0 +1,865 @@
+
+/obj/effect/warped_rune/adamantinespace  //doesn't have a rune icon yet please spriters help me I can't sprite for shit I beg you
+	desc = "The universe's ressource are nothing but tools for us to use and abuse."
+	max_cooldown = 300  //"mines" things every 30 seconds.
```
generally, this is a pretty shitty rune. by the time you get this, the miners should have already mined all of the ores you'll ever need. if they haven't, just grabbing a pickaxe, ore bag, and mining scanner from the public part of the mining base and mining ore manually is going to be less effort than making and setting one of these things up and then going back and harvesting a mere 20 chunks of ore from it 10 minutes later, hell, instead of making this crossbreed from an adamantine slime, you could just make a servant golem shell from it and tell your new servant golem to go mine mats for you.

ATH1909 replied:
tldr;
this rune is way too slow, requires too much effort/costs too much to make for what it does, and the thing that it's supposed to do for you should have been done an HOUR ago by your miners.

Caldony replied:
The main point of this rune is to not only act as free mining point, but also to replace miners if they somehow didn't do their job and potentially getting a lot of the rarer ores. While it can't per say replace the speed of actual miners it does something that doesn't require any effort apart from just placing it and getting it later.

Caldony committed:
Updates rune sprites slightly

gives an actual sprite for the adamantine rune and use rewind to cut on some useless sprite animation.
My sprites are still god awful though sorry

Caldony committed:
minor convenience tweaks

Gold will now turn things to gold on a cooldown and keep in mind the actual material worth of the item.
shitty sprite for adamantine. 
red rune now gives less punch damage but can get more than one person at a time and don't deal stamina damage to its user(still compatible with gloves of the north star)
removed a few useless command
crossbreed extract will now tell you how many charges they have left
cerulean now has 6 charges instead of 1
and probably a few other things I forgot to mention

Caldony committed:
minor tweaks

more crossbreed works using crossed() instead of signals
few comment changes.

Caldony committed:
add light pink (no unique sprite for the rune yet)

Adds the light pink crossbreed!
this extract can create a rune capable of reviving people as long as they have a brain and their soul isn't capable of coming back to their body. in which case it will fully heal them and pull in a random ghost to take their place! Not a perfect alternative to actual defibrillation considering you need the soul gone for it to work and a willing ghost.

Caldony committed:
new rune sprites

new (really shitty) rune sprites for the light pink rune and the black rune

Caldony committed:
adds the black crossbreed!

the black crossbreed creates the black rune, a rune capable of swapping the species of the first two people that walk on the rune, works with any human subset such as abductors, golems, nightmares and anything else under that subset. Also, give a proper "sprite" (coder sprite) to the light pink rune.

Caldony committed:
minor tweaks

moocowswag commented:
Does the red rune work with telekinesis?

I can only imagine how hard it would be to fight that.

Caldony commented:
@moocowswag Since it directly changes species punch damage it probably works with the extendo hand, The red rune is the hardest to balance right now because it has a lot of variables such as gloves of the north star, extendo hand, martial arts (since the damage are additive and not set) and tons of other things. I can't say for sure how I'm going to balance it, or if the rune won't have to be reworked from the ground up.

Rohesie commented:
The new file should be `.dm` and added to the `.dme`

Caldony committed:
rename warping to warping.dm

Caldony commented:
@Rohesie The file has been renamed, I don't know how I didn't see this earlier.

Rohesie commented:
You haven't tested this PR, have you?

Caldony commented:
@Rohesie I tested the PR multiple times and it worked as intended for each crossbreed with no bugs or runtime that I could find. warping.dm was effectively the only file that I didn't upload directly and instead copy-pasted the original code unto, which was a pretty stupid thing to do in retrospect.

Caldony committed:
Links the warping file to the .dme

I thought the files were linked automatically. They weren't

Caldony committed:
Fixes Crossed() not calling their parent

Rohesie commented:
It's fine, you tested an earlier version.

The process of reviewing can be taxing for both the person doing and the person receiving the review because as things change there's need for further testing and that can get tedious. But it's normal.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	move_resist = INFINITY //here to avoid the rune being moved since it only sets it's turf once when it's drawn. doesn't include admin fuckery.
+
+///runes can also be deleted by bluespace crystals if the xenobiologist is fucking everyone up and you need a way to destroy the rune.
+/obj/effect/warped_rune/attackby(obj/item/stack/BC, mob/user)
```
Try using more informative names than `BC`

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	desc = "An unstable rune born of the depths of bluespace"
+	icon = 'icons/obj/slimecrossing.dmi'
+	icon_state = "greyspace_rune"
+	var/runepath = /obj/effect/warped_rune  ///serves as a way for some specific runes to identify themselves in the turf.TODO: you can just use src dumbass
```
Documentation should go on the line over the variable.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	var/storing_time = 5 ///is only used for bluespace crystal erasing as of now
+	anchored = TRUE
+	layer = MID_TURF_LAYER
+	resistance_flags = FIRE_PROOF  //It's only fireproof because of the fire rune and because how do you burn a rune anyway.
```
Put the object-specific variables (the ones that start with `var/` at the bottom and the redefined ones (the ones without) at the top, for clarity.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	anchored = TRUE
+	layer = MID_TURF_LAYER
+	resistance_flags = FIRE_PROOF  //It's only fireproof because of the fire rune and because how do you burn a rune anyway.
+	var/turf/T //used on a LOT of runes, so we might as well put it here
```
Try using better names than one or two uppercase letters for vars.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	qdel(src)
+
+ ///using the extract on the floor will "draw" the rune.
+/obj/item/slimecross/warping/afterattack(turf/target, mob/user, proximity)
```
Keep the type of the input as the possible ones, not just the one you want.
Other coders might want to change the proc and this can be misleading.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	name = "greyspace rune"
+	desc = "Death is merely a setback, anything can be rebuilt given the right components"
+	icon_state = "greyspace_rune"
+	var/i = 0 //number of slime extract currently absorbed by the rune
```
This is a very uninformative var name.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	desc = "Death is merely a setback, anything can be rebuilt given the right components"
+	icon_state = "greyspace_rune"
+	var/i = 0 //number of slime extract currently absorbed by the rune
+	var/mob/living/simple_animal/slime/S //S stands for the slime that will be spawned
```
This one as well, try to improve the var names.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///Makes a slime of the color of extract that were put on the runes.can only take one type of extract between slime spawning.
+/obj/effect/warped_rune/greyspace/proc/slimerevival()
+	if(i <! 8) // this shouldn't happen and will bring back the count to 7
```
Typo?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+			return
+		if(P.cell.charge <! (P.cell.maxcharge))
+			return
+		P.cell.charge += P.cell.maxcharge/10 //will basically recharge 10% of the APC cell every 10 seconds.
```
```suggestion
		P.cell.charge += P.cell.maxcharge * 0.1 //will basically recharge 10% of the APC cell every 10 seconds.
```
Prefer to use multiplications over divisions.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+				C.update_icon()
+				if(C.charge > C.maxcharge)
+					C.charge = C.maxcharge
+					return
```
Try a redefined proc instead of a long `istype()` chain.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+/* Dark purple crossbreed, Fill up any beaker like container with 50 unit of plasma dust every 30 seconds  */
+
+
```
Is there need for these big blank sections?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///the first two person that stepped on the rune swap places after the second person stepped on it. TODO: fix the potential lag abyse
+obj/effect/warped_rune/bluespace/Crossed(atom/movable/AM)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///stops time on a single tile for 5 seconds
+obj/effect/warped_rune/sepiaspace/Crossed()
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///makes a hologram of the mob stepping on the tile, any new person stepping in will replace it with a new hologram
+/obj/effect/warped_rune/ceruleanspace/Crossed(atom/movable/AM)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+///destroys the hologram with the rune
+/obj/effect/warped_rune/ceruleanspace/Destroy()
+	..() //needed so the rune still gets stored by the crossbreed
+	qdel(holotile)
```
```suggestion
/obj/effect/warped_rune/ceruleanspace/Destroy()
	qdel(holotile)
	return ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///colors whoever steps on the rune randomly
+/obj/effect/warped_rune/pyritespace/Crossed(atom/movable/AM)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///boost up the unarmed damage of the person currently on the tile.
+/obj/effect/warped_rune/redspace/Crossed(atom/movable/AM)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	for(var/mob/living/carbon/human/D in T) // takes away people's punch damage that are on the rune anyway.
+		H.dna.species.punchstunthreshold -= 20
+		H.dna.species.punchdamagelow -= 20
+		H.dna.species.punchdamagehigh -= 20.
```
Always `return ..()` at the end of `Destroy()` procs.
Preserve the return and do it at the end of the proc.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+
+///adds the jolly mood effect along with hug sound effect.
+/obj/effect/warped_rune/pinkspace/Crossed(atom/movable/AM)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	
+///will swap the species of the first two human or human subset that walk on the rune	
+/obj/effect/warped_rune/blackspace/Crossed(atom/movable/AM)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/research/xenobiology/xenobiology.dm:
```diff
@@ -17,7 +17,7 @@
 	var/recurring = FALSE
 	var/research ///Research point value for slime cores. These are defines stored in [/__DEFINES/research] - the actual values are updated there.
-
+	var/color_slime ///the color of the extract and the slime it came from
```
How about a type path instead of a string?

Caldony replied:
What do you mean by a type path? I used a string because the slime subtype use strings in their list to chose their color so I thought it was more appropriate.

Rohesie replied:
Yea, nevermind, this makes sense for slimes.

Caldony committed:
renamed variable and bug fixing.

the yellow core now uses normal processing instead of machine processing to avoid runtime
the grey rune now use the extract type directly instead of its color value, making the xenobiology.dm changes useless, I will revert its changes after this commit
most variable names have been changed to be more straight forward and clear.
auto_doc regulation problem have been taken into account concerning variables
thanks to Rohesie for bringing up these issues

caluany committed:
removes unneeded variables

Xenobiology does not need to be changed as the warped grey crossbreed can create a slime with just the extract type.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -10,39 +10,53 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
-	var/max_charge = 1 ///max number of charge, might be different depending on the crossbreed (all crossbreed have 1 max charge for now)
-	var/storing_time = 15 ///time it takes to store the rune back into the crossbreed
-	var/drawing_time = 15 ///time it takes to draw the rune
+	///default color of the crossbreed
+	colour = "grey" 
```
No need to document this var here. Do it at the variable definition instead.

caluany committed:
Merge branch 'master' of https://github.com/Caldony/tgstation

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -10,39 +10,53 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
 	icon_state = "greyspace_rune"
-	var/runepath = /obj/effect/warped_rune  ///serves as a way for some specific runes to identify themselves in the turf.TODO: you can just use src dumbass
-	var/storing_time = 5 ///is only used for bluespace crystal erasing as of now
+	///here to avoid the rune being moved since it only sets it's turf once when it's drawn. doesn't include admin fuckery.
+	move_resist = INFINITY 
+	///serves as a way for some specific runes to identify themselves in the turf. TODO : use src instead
```
Same here, these are comments, not variable documentation. Put them back to old format: with just `//` after the variable.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -10,39 +10,53 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
+	///is only used for bluespace crystal erasing as of now
+	var/storing_time = 5 
+	///Nearly all runes needs to know which turf they are on
+	var/turf/T 
```
Better var name please.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -129,27 +144,27 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
 /obj/effect/warped_rune/greyspace/proc/slimerevival()
-	if(i <! 8) // this shouldn't happen and will bring back the count to 7
-		i = 7
+	if(absorbed_extract <! 8) // this shouldn't happen and will bring back the count to 7
```
Typo?

caluany committed:
Revert "Merge branch 'master' of https://github.com/Caldony/tgstation"

This reverts commit 5b66b0c55948b434ab4b6701415414ac05ca9616, reversing
changes made to bb646e3f675fe451ec871929a54dd36aeafca0c9.

caluany committed:
Revert "removes unneeded variables"

This reverts commit bb646e3f675fe451ec871929a54dd36aeafca0c9.

caluany committed:
Revert "fix two color_slime variable missing quotes"

This reverts commit 40f78649acd9b552d68a84481df369960a6009e2.

caluany committed:
Revert "Update xenobiology.dm"

This reverts commit 70c252c85328a1d2cecb0e1e43aeeca6c6b0d371.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,918 @@
+	))
+
+/obj/item/stack/sheet/xeno_resin/get_main_recipes()
+	.= ..()
+	. += GLOB.resin_recipes
+
+
+
+
+
+
+
+/* pink rune, makes people slightly happier after walking on it*/
+
+/obj/item/slimecross/warping/pink
```
The formatting on this PR is quite an eye sore. Try improving it some, use some kind of format and stick to it. Be consistent.
If you have none try using spaces between the operators (`. = ..()` instead of mixed cases `.= ..()` or `.=..()`), and put one or two spaces between each object or proc definition instead of a random amount.

Caldony committed:
fixes variables and bugs

I already did this but I don't know how to use github desktop and I reverted it
in short : better variable names, yellow rune no longer runtime, crossed() now use .=..()

caluany committed:
Revert "Revert "Update xenobiology.dm""

This reverts commit 2c363356635e377b127bb9300a5e2a155837acbb.

caluany committed:
Revert "Revert "fix two color_slime variable missing quotes""

This reverts commit 74dd4f87bad8576f2d5c1a56f4d91bf502073fd9.

Caldony committed:
Code formatting

I try to format the code in a more appropriate way by :
putting two empty space between each proc/object
putting one empty space after an early return.
replacing .=..() with . = ..() to stay consistent
renaming a few variables (again) such as T being renamed rune_turf and AM for the crossed() proc as crossing.
And probably a few other things I forgot to mention.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -1,54 +1,52 @@
-	
-	var/cooldown = 0 
+	var/turf/rune_turf
+	var/cooldown = 0
```
Please document all new variables.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -58,83 +56,85 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
 		return
 
-	if(istype(target) && locate(/obj/effect/warped_rune) in target) //check if the target is a floor and if there's a rune on said floor
+	if(istype(target,/turf) && locate(/obj/effect/warped_rune) in target) //check if the target is a floor and if there's a rune on said floor
```
```suggestion
	if(isturf(target) && locate(/obj/effect/warped_rune) in target) //check if the target is a floor and if there's a rune on said floor
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -58,83 +56,85 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
 			return
 
-	if(!istype(target) || isspaceturf(target))
+	if(!istype(target,/turf) || isspaceturf(target))
```
```suggestion
	if(!isturf(turf) || isspaceturf(target))
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -1,54 +1,52 @@
 ///runes can also be deleted by bluespace crystals if the xenobiologist is fucking everyone up and you need a way to destroy the rune really fast.
-/obj/effect/warped_rune/attackby(obj/item/stack/space_crystal, mob/user)
-	if(!istype(space_crystal,/obj/item/stack/sheet/bluespace_crystal) && !istype(space_crystal,/obj/item/stack/ore/bluespace_crystal))
+/obj/effect/warped_rune/attackby(used_item, user)
```
Are users no longer guaranteed to be mobs?
Or used items no longer be guaranteed to be items?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -58,83 +56,85 @@ put up a rune with bluespace effects, lots of those runes are fluff or act as a
-/obj/item/slimecross/warping/afterattack(turf/target, mob/user, proximity)
+
+///using the extract on the floor will "draw" the rune.
+/obj/item/slimecross/warping/afterattack(target, user, proximity)
```
Don't remove the var types if they are guaranteed to be of those types when called.

Caldony committed:
code tweaks, variable documenting

more variables are now documented correctly
the var runepath of the parent rune has been deleted as it's been rendered obsolete with recent changes (not the var runepath of the warping extract.)
readds proper typing to the attacked proc of the extract/rune

Caldony committed:
Improves yellow rune code

did you know that get_cell() is a proc that exist?
the yellow rune now only use one proc for recharging items instead of checking the type of each item on it.
In short, it's much easier on the eye and more efficient, but the rune is functionally the same.

Caldony committed:
correctly documents variable of xenobiology.dm

just puts the /// above the variable in question

Caldony committed:
change the warping afterattack() parameter

Change the warping afterattack() parameter from turf/target to atom/target, functionally the code is the same but more convenient to work with.

Caldony committed:
Reworks the rune drawing proc

reworks the after_attack() proc for the warping extract so it doesn't work on lava and chasm

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+
+
+///runes can also be deleted by bluespace crystals if the xenobiologist is fucking everyone up and you need a way to destroy the rune really fast.
+/obj/effect/warped_rune/attackby(obj/item/stack/used_item, mob/user)
```
```suggestion
/obj/effect/warped_rune/attackby(obj/item/used_item, mob/user)
```
Keep the incoming var correctly typed. Type cast it if needed, but don't give it a type it might not necessarily have on the proc redefinition.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+			to_chat(user, "<span class='notice'>You store the rune in [src].</span>")
+			qdel(target)
+			warp_charge++
+			desc = "It just won't stay in place. it has [src.warp_charge] charge left"
```
```suggestion
			desc = "It just won't stay in place. it has [warp_charge] charge left"
```
Only use `src.` if there are two variables with the same name, one defined for the proc, one for the source.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+		to_chat(user, "<span class='warning'>You can't draw the rune here!</span>")
+		return
+
+	if(warp_charge > 0 && do_after(user, drawing_time,target = target)) //spawns the right rune if you have charge(s) left
+		if(warp_charge <= 0) //this is only here to fix a bug where the user can draw multiple runes at once if they spam click
+			to_chat(user, "<span class='warning'>[src] is empty!</span>")
+			return
```
```suggestion
	if(LAZYLEN(user.do_afters))
		return //To prevent click-spam.
	if(warp_charge <= 0)
		to_chat(user, "<span class='warning'>[src] is empty!</span>")
		return
	if(do_after(user, drawing_time,target = target)) //spawns the right rune if you have charge(s) left
		if(warp_charge <= 0) //In case the state has changed since we started the do_after
			to_chat(user, "<span class='warning'>[src] is empty!</span>")
			return
```
Early returns and whatnot.
You could also check `target.targeted_by` if you want for only one thing to manipulate it at a time.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+
+		playsound(target, 'sound/effects/slosh.ogg', 20, TRUE)
+		warp_charge--
+		desc = "It just won't stay in place. it has [src.warp_charge] charge left"
```
```suggestion
		desc = "It just won't stay in place. it has [warp_charge] charge left"
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+	icon_state = "greyspace_rune"
+	max_cooldown = 50
+	///number of slime extract currently absorbed by the rune
+	var/absorbed_extract = 0
```
`absorbed_extracts_num` would be a bit more intuitive, or just `absorbed_extracts`

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+	var/absorbed_extract = 0
+	///mob path of the slime spawned by the rune
+	var/mob/living/simple_animal/slime/spawned_slime
+	///extractype is used to remember the type of the extract on the rune
+	var/extractype = 0
```
Is this a numeric var? Why the zero?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,909 @@
+	///number of slime extract currently absorbed by the rune
+	var/absorbed_extract = 0
+	///mob path of the slime spawned by the rune
+	var/mob/living/simple_animal/slime/spawned_slime
```
This needs to be cleared properly on both the rune and the slime's `Destroy()` for garbage collection.

Caldony replied:
I'm not sure I understand what you mean, should I change the variables to null when the rune is destroyed?

Rohesie replied:
Yes, and also when the slime is destroyed.
Not all variables, just the ones that reference the object.

Caldony committed:
adds the warped_room template

used for the rainbow place

Caldony committed:
adds "warped_room" to the areas.dmi

Caldony committed:
adds oil rune sprites and rainbow rune sprite

I still can't sprite please help

Caldony committed:
adds rainbow and oil runes

Caldony committed:
Adds warping rainbow and oil

rainbow will teleport you into a unique room like hilbert's hotel
oil will protect people on the rune from outside explosions.

Caldony committed:
map didn't merge correctly

Caldony committed:
fixes the new area icon

caluany committed:
Revert "fixes the new area icon"

This reverts commit 50086a299b190b20c3e1286bf982958461c56f0d.

Caldony committed:
add new line

I should really stop testing things only locally

Caldony committed:
Adds a proper area to the rainbow rune room

let's hope it doesn't break this time

caluany committed:
Revert "Adds a proper area to the rainbow rune room"

This reverts commit 2fa4e553da9400ac0af50a8130834a300838859f.

Caldony committed:
adds the warped_rune area """sprite"""

caluany committed:
Revert "adds the warped_rune area """sprite""""

This reverts commit 9969a9c809fa72d0874034b9bb9945204806d5f0.

Caldony committed:
why isn't the sprite working I swear to god

caluany committed:
Revert "why isn't the sprite working I swear to god"

This reverts commit c3f81cd3f8c6810706ae0356310f519fa1d95f0d.

caluany committed:
Revert "adds "warped_room" to the areas.dmi"

This reverts commit df70d36561f85e80c605c263dec38a8df1a624ba.

Caldony committed:
fuck it I'll just change the area sprite to yellow

* change the area sprite to yellow

* adds the area properly

Caldony commented:
I'm sorry for the mess you just witnessed, I tried to fix problematic new sprites in the area.dmi file (that worked just fine when i tested it mind you) and I had a panic attack and made some questionable choices. Sorry for the maintainer who's going to have to wake up to that shit. Either way the area is functionnal it just uses yellow to differentiate itself

Caldony commented:
I'll be closing this PR now, while it hasn't been marked as stale and I did a fair amount of progress since the start I simply don't understand github and the PR process well enough to recommend merging this in good faith. I may reopen a newer version of this later down the line once I feel confident enough in my coding abilities and I consider it "finished" instead of making stupid decisions in a fit of panic. Thank you for all the criticism that helped me improve this code. If you want to test it for yourself all of it is functional with minimum runtime on the last version of TG I tested it on.
